File: framework/languages/dotnet/perf/ZLink.Framework.Perf.Shared/MetricCatalog.cs
using System.Text.Json;

namespace ZLink.Framework.Perf;

public static class MetricCatalog
{
    private static readonly JsonDocument Catalog = JsonDocument.Parse(File.ReadAllText(Path.Combine(AppContext.BaseDirectory, "metric-catalog.json")));
    private static string[] Strings(string name) => Catalog.RootElement.GetProperty(name).EnumerateArray().Select(value => value.GetString()!).ToArray();
    public static readonly string[] LatencySuffixes = Strings("latencySuffixes");
    public static readonly string[] Outcomes = Strings("outcomes");
    private static readonly string[] Scalars = Strings("counts").Concat(Strings("rates")).Concat(Strings("unsupported"))
        .Concat(Strings("latencyPrefixes").SelectMany(prefix => LatencySuffixes.Select(suffix => prefix + "." + suffix))).ToArray();
    private static readonly string[] Histograms = Catalog.RootElement.GetProperty("histogramPrefixes").EnumerateObject().Select(property => property.Name).ToArray();
    public static void Null(Dictionary<string, object?> values, Dictionary<string, NullReason> reasons,
        string container, string key, string code, string reason)
    {
        values[key] = null;
        reasons["/" + container + "/" + key] = new(code, reason);
    }
    public static void BaselineNulls(Dictionary<string, object?> metrics, Dictionary<string, object?> histograms,
        Dictionary<string, NullReason> reasons)
    {
        foreach (var key in Scalars) Null(metrics, reasons, "metrics", key, "NOT_APPLICABLE", "This scenario does not own this observation.");
        foreach (var key in Histograms) Null(histograms, reasons, "histograms", key, "NOT_APPLICABLE", "This scenario does not measure this interval.");
        foreach (var suffix in new[] { "p50Ms", "p95Ms", "p99Ms" })
            Null(metrics, reasons, "metrics", "host.queueWaitLatency." + suffix,
                "PUBLIC_OBSERVATION_UNSUPPORTED", "Public status provides no exact pre-receive to handler queue-wait hook.");
    }
}

File: framework/languages/java/perf/src/main/java/systems/zlink/perf/MetricCatalog.java
package systems.zlink.perf;

import com.fasterxml.jackson.databind.JsonNode;
import java.io.IOException;
import java.util.Map;

/** Shared result keys have one owner in framework/perf-contract. */
public final class MetricCatalog {
    private static final JsonNode CATALOG;
    static {
        try (var input=MetricCatalog.class.getResourceAsStream("/metric-catalog.json")) {
            if(input==null)throw new IllegalStateException("Shared metric catalog is missing");
            CATALOG=Config.JSON.readTree(input);
        }catch(IOException error){throw new ExceptionInInitializerError(error);}
    }
    private MetricCatalog(){}
    public static void complete(Map<String,Object> metrics,Map<String,Object> histograms,Map<String,Object> reasons){
        for(String group:new String[]{"counts","rates","unsupported"})for(JsonNode key:CATALOG.path(group))
            missing(metrics,reasons,"metrics",key.asText(),group.equals("unsupported")?"PUBLIC_OBSERVATION_UNSUPPORTED":"NOT_APPLICABLE");
        for(JsonNode prefix:CATALOG.path("latencyPrefixes"))for(JsonNode suffix:CATALOG.path("latencySuffixes"))
            missing(metrics,reasons,"metrics",prefix.asText()+"."+suffix.asText(),"NOT_APPLICABLE");
        CATALOG.path("histogramPrefixes").fieldNames().forEachRemaining(key->missing(histograms,reasons,"histograms",key,"NOT_APPLICABLE"));
        for(JsonNode key:CATALOG.path("errorNamespaces"))metrics.putIfAbsent(key.asText(),Map.of());
    }
    private static void missing(Map<String,Object> values,Map<String,Object> reasons,String container,String key,String code){
        if(values.containsKey(key))return;
        values.put(key,null);reasons.put("/"+container+"/"+key,Measurement.reason(code,"This role has no applicable public observation"));
    }
}
